v3/lints/cabf_br/lint_subject_common_name_not_exactly_from_san.go

```diff
@@ -0,0 +1,69 @@
+package cabf_br
+
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type subjectCommonNameNotExactlyFromSAN struct{}
+
+/************************************************
+If present, this field MUST contain exactly one entry that is one of the values contained
+in the Certificate's `subjectAltName` extension
+
+If the [subject:commonName] is a Fully-Qualified Domain Name or Wildcard Domain Name, then
+the value MUST be encoded as a character-for-character copy of the dNSName entry value from
+the subjectAltName extension.
+************************************************/
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_subject_common_name_not_exactly_from_san",
+		Description:   "The common name field in subscriber certificates must include only names from the SAN extension",
+		Citation:      "BRs: 7.1.4.2.2",
+		Source:        lint.CABFBaselineRequirements,
+		EffectiveDate: util.CABFBRs_1_8_0_Date,
+		Lint:          NewSubjectCommonNameNotExactlyFromSAN,
+	})
+}
+
+func NewSubjectCommonNameNotExactlyFromSAN() lint.LintInterface {
+	return &subjectCommonNameNotExactlyFromSAN{}
+}
+
+func (l *subjectCommonNameNotExactlyFromSAN) CheckApplies(c *x509.Certificate) bool {
+	return c.Subject.CommonName != "" && !util.IsCACert(c)
```

Comment:
Design Question: Should we instead be checking if `len(c.Subject.CommonNames) > 0` here?

Context:
1. What's the expected behaviour if I supply an empty commonName (a zero-length string as the value, but a CN attribute present in the DN). My reading is that this will return False, and the Lint will succeed. We have `cabf_br/lint_subject_common_name_included` (as a notice for using the deprecated field), but AFAICT, no other lint will fire here to signal that the empty string is an error.
2. In the `Execute` call, my understanding is that we'll only test the 'last' commonName. We have `cabf_br/lint_extra_subject_common_names` to test if there are multiple CNs, so this is more of a design question about whether we should run this check on every commonName present or not.

Reply:
1. If you supply an empty common name that is equivalent to not providing one at all.  In the zcrypto/x509/Certificate struct the CommonName is a string type which in go doesn't have a nil value.  The closest thing to one is the empty string, if it were a *string then it could be nil but as it isn't then a certificate with no common name or a zero length common name would result in the same thing as far as this lint is concerned.
2. That is correct, this lint checks only the commonName field which I believe does check only the final commonName provided.  I wasn't aware that it was possible for a certificate to have more than one to be honest but if it is deemed to be valid behaviour it's an easy change to make sure they're all present.

Reply:
1. Correct, that’s why checking `len(c.Subject.CommonNames) > 0` allows you to distinguish absent-vs-empty
2. Yeah, it’s valid in X.509, invalid in the BRs. While we have another lint that checks there is only one CN, and even though the same section (of SC48) makes it clear there’s one CN, I _think_ the approach has been to treat the lints independently - i.e. check all CNs here.

Reply:
Unless, I've missed your meaning @sleevi?  When you say actively providing an empty CN then would that be considered valid or not?  I guess one could say that if there is nothing in the CN field then there should be a SAN field with nothing in it to match but that would violate the requirement for a SAN dNSName to contain either a FQDN or a Wildcard domain so this does feel like it's getting a bit academic...

I guess the thing to add is to sort of repeat my earlier point and say that with zcrypto/x509 as it is now then it may not be possible to tell the difference between a parsed certificate with no CN and one with an empty CN when it comes to looking at them both in zlint.

Reply:
@robplee Supplying an empty CN is invalid, and this would be the one lint responsible for detecting that.

I’m not sure the point your making about ZCrypto - I’m saying it is possible. Note I’m saying `CommonNames` - plural. There’s the singular field that just contains the last value, and then the plural field that is a slice that contains all values. If the slice has any value, then we know there was at least one CN present.

Reply:
Just wanted to point out that we might get empty CN attribute value checking “for free” by the ASN.1 decode logic, as the ASN.1 CN attribute type has a constraint of least one character.

Reply:
Apologies, I replied to myself without seeing your reply.

I will make some test data and see what happens but I think it'll behave as we expect.  From looking at the zcrypto code I think it _should_ keep adding empty common names into the array of strings so I think this is the right approach for catching this issue.  I'll make some test data to confirm.

However, I think it might be better for empty CNs to get their own lint as currently zlint will return an error declaring "The common name field in subscriber certificates must include only names from the SAN extension" and saying the e_subject_common_name_not_exactly_from_san lint has failed if there is an empty CN which doesn't seem quite right!

One quick question: what would the effective date be of such a lint?  BR 1.8.0 date or earlier?

Reply:
@CBonnell Zlint doesn’t do a lot of the ASN.1 constraint checks. AFAICT, there’s no lint today for empty CN.

@robplee An empty (zero-length) CN should be seen as “a value”; that is, it is semantically distinct than omitting CN. As @CBonnell points out, it’s invalid ASN.1, but it’s something easy to address in this lint.

That is, you get this “for free” if you just check all of the CNs in `c.Subject.CommonNames`, which is why I was flagging it. That’ll handle the empty/invalid ASN.1 case (at least, to the extent there’s not an empty SAN). So just changing this to check all the names will address this - sorry for not making that clearer.

Reply:
After battling to get some test cases to confirm I've extended the lint such that it applies based on the len(CommonNames) and that does give us empty CN catching for free so I think we're slowly getting this PR closer to where we want it.